### ISSUE TITLE
Start a Script after connection / before disconnection?

### ISSUE THREAD

**Issue by mariolorenz** [jochenwierum/openvpn-manager#46]
Dear Sir,

can i start a script automaticly after connection and before disconnection?
What should i do?

Thanks!

**Comment by jochenwierum** [comment]
Hi mario,

you can do this with openvpn itself. In your configuration file, you can start custom programs in various situations. They can be registered using the following directives:

```
# your normal .ovpn content
# ...

up c:\my\command
down-pre c:\my\other\command
down c:\my\other\command
```

The behavior can be customized by using e.g. `up-delay`, `ping`, `up-restart`. The [manual of openvpn](https://openvpn.net/index.php/open-source/documentation/manuals/65-openvpn-20x-manpage.html) describes the directive in more detail.

Cheers,
Jochen

**Comment by mariolorenz** [comment]
Hello,

do you have experience with "up" and "down" on Windows 10? These are, according to documentation, exactly the options that I need. Thank you.

I included them and set script-security to "2". This is important for the contents in the bat-files.

The "up"-bat-file containing drive mount commands. (e.g. net use m: \\server\path)

If I start a connection now, in the OVPN-Manager-Window the log is "frozen" and stays open for 4 minutes. After the break, the connection would be closed. 

In the log is written:
Wed Jan 24 19:11:42 2018 WARNING: Failed running command (--up/--down): returned error code 1
Wed Jan 24 19:11:42 2018 Exiting due to fatal error

When I run the scripts by hand, then everything works fine. The opvn config files also work smoothly without "up" and "down". Does it have something to do with the administrator privileges? Do I have to give any component administrator rights?

Regards,

**Comment by hapm** [comment]
Is openvpn running as a service or in user mode?

**Comment by mariolorenz** [comment]
I run openVPN as service. But i can change it, if it is the solution. I can test it in the afternoon ...

**Comment by mariolorenz** [comment]
Hello again,
i´ve test something. Also with OpenVPN Gui. In OpenVPN Gui the same phenomenon occurs. So it has nothing to do with OpenVPN Manager. So I  close this comment.

**Comment by Jakobud** [comment]
What was the solution to the `Failed running command (--up/--down): returned error code 1` problem? I am running OpenVPN in Windows 10 in user mode as Administrator and I'm getting this error when trying to execute some .bat files in the `up` and `down` directives.

**Comment by hapm** [comment]
Please try to run your configuration with openvpn or openvpn gui. If the error still happens it's probably a problem with the batch script itself. Another problem could be, that you need to escape special characters like \ in the arguments if the directives.

**Comment by jochenwierum** [comment]
In addition:
Did you specify "script-security 2"in your ovpn file?

See https://openvpn.net/vpn-server-resources/explanation-of-client-side-scripting-with-simple-examples/ for further explainations.

Cheers
Jochen

**Comment by Jakobud** [comment]
> Please try to run your configuration with openvpn or openvpn gui

Yes I'm running it in Windows 10 with openvpn gui (the one where you right click the icon in the notification icon area in order to connect/disconnect).

Yes I have `script-security 2` in my config. This is a sample of it:

```
...
script-security 2
up up.bat
down down.bat
...
```

up.bat:
```
ROUTE ADD 0.0.0.0 MASK 0.0.0.0 %ifconfig_remote%
```

down.bat:
```
ROUTE DELETE 0.0.0.0 MASK 0.0.0.0 %ifconfig_remote%
```

Nothing really crazy going on in the batch files. They require elevated permissions obviously to run but since OpenVPN is already running as Administrator I would assume that any scripts run via the config would also use admin privileges.

OpenVPN just gives the error same as above: `WARNING: Failed running command (--up/--down): returned error code 1`.

If I start up a command prompt with administrative privileges and manually execute the batch files they run just fine.

Also, if this helps, the above up/down batch files scripts run just fine in Windows 7 via the same OpenVPN config.

Also I've tried changing the `up up.bat` and `down down.bat` in the config to be the full path of the `up.bat` and `down.bat` with the same result.

**Comment by hapm** [comment]
Wouldn't it be easier to use the [`route`](https://openvpn.net/community-resources/setting-up-routing/) directive for settings routes? Anyway:

* did you test the config directly with openvpn and if so does it work? 
* did you try to use full pathes to the batch files?

**Comment by Jakobud** [comment]
Hmmm. Can you explain how I would recreate the above `ROUTE` commands using the `route` directive in the config? If that works that would be a desirable option. I'm fairly new to OpenVPN and I'm not familiar with some of this. Would it just be this:

```
route 0.0.0.0 0.0.0.0
```

? Essentially, my route I'm trying to create sets up the VPN so that Windows does not use the VPN connection when it's active. This is because I only want certain applications to use that VPN connection (this part is configured in various programs where you can choose which adapter to use). But not everything. Using the above `ROUTE` batch files accomplishes this. It makes it so that when the VPN connection is made, all of Windows ignores it by default and that it continues just using your normal ISP connection.

What do you mean `directly with openvpn`? Do you mean just executing OpenVPN from the commandline instead of the GUI?

I did try full paths to the batch files in the config.

**Comment by hapm** [comment]
If your goal is to not set a default route remove `redirect-gateway` from your client config and make sure no route is pushed by the server.

**Comment by Jakobud** [comment]
I can't seem to get this to work. I don't have `redirect-gateway` in my config. Here is my config:

```
client
dev tun
proto udp
remote us3.privacy.network 1198
resolv-retry infinite
nobind
persist-key
persist-tun
cipher aes-128-cbc
auth sha1
tls-client
remote-cert-tls server
block-outside-dns
auth-user-pass auth
compress
verb 1
reneg-sec 0
<crl-verify>
-----BEGIN X509 CRL-----
.....................................
-----END X509 CRL-----
</crl-verify>

<ca>
-----BEGIN CERTIFICATE-----
.................................
-----END CERTIFICATE-----
</ca>

disable-occ
```

So, as expected, when I connect, this makes all my internet traffic use my VPN IP address. I'm trying to avoid this. I want my internet traffic to continue using my ISP-assigned IP address, just as if I was not connected to the VPN.

Then, I would have certain software set to use my VPN adapter connection so that only that software is using the VPN IP address   while the rest of Windows is still using my ISP IP address.

Using that above batch files solved this (`ROUTE ADD 0.0.0.0 MASK 0.0.0.0 %ifconfig_remote%` and `ROUTE DELETE 0.0.0.0 MASK 0.0.0.0 %ifconfig_remote%`). But I can't seem to recreate this using only the OpenVPN configuration.

I also tried adding `route 0.0.0.0 0.0.0.0` to my config and that didn't seem to have any effect.
